src/core/libmaven/Scan.cpp

```diff
@@ -539,7 +537,72 @@ vector<int> Scan::assignCharges(MassCutoff *massCutoffTolr) {
     return parentPeaks;
 }
 
-double Scan::getPrecursorPurity(float ppm = 10.0) {
-    // Just a stub for now. See Eugene's maven_core for actual implementation.
-    return 0.0;
+Scan* Scan::getLastFullScan(int historySize)
+{
+	if (!this->sample) return 0;
+    int scanNum = this->scannum;
+    for(int i = scanNum; i > (scanNum - historySize); i--) {
+        Scan* lscan = this->sample->getScan(i);
+        if (!lscan or lscan->mslevel > 1) continue;
+		return lscan; // found ms1 scan, all is good
+	}
+	return 0;
+}
+
+vector<mzPoint> Scan::getIsolatedRegion(float isolationWindowAmu)
+{
+	vector<mzPoint> isolatedSegment;
+	if(! this->sample) return isolatedSegment;
+
+	//find last ms1 scan or get out
+	Scan* lastFullScan = this->getLastFullScan();
+	if (!lastFullScan) return isolatedSegment;
+
+	//no precursor information
+	if (this->precursorMz <= 0) return isolatedSegment;
+
+	//extract isolated region 
+	float minMz = this->precursorMz - (isolationWindowAmu / 2.0);
+	float maxMz = this->precursorMz + (isolationWindowAmu / 2.0);
+	
+	for(int i = 0; i < lastFullScan->nobs(); i++ ) {
+		if (lastFullScan->mz[i] < minMz) continue;
+		if (lastFullScan->mz[i] > maxMz) break;
+		isolatedSegment.push_back(mzPoint(lastFullScan->rt,
+                                          lastFullScan->intensity[i],
+                                          lastFullScan->mz[i]));
+	}
+	return isolatedSegment;
+}
+
+double Scan::getPrecursorPurity(float ppm)
+{
+    if (this->precursorMz <= 0 ) return 0;
+    if (this->sample == 0 ) return 0;
+
+	//extract isolated window
+    vector<mzPoint> isolatedSegment = this->getIsolatedRegion(this->isolationWindow);
+	if (isolatedSegment.size() == 0) return 0;
+
+	//get last full scan
+	Scan* lastFullScan = this->getLastFullScan();
+	if (!lastFullScan) return 0;
+
+	//locate intensity of isolated mass
+    MassCutoff* massCutoff = new MassCutoff();
```

Comment:
When done with `massCutoff` lets deallocate it? Although it probably does not take up a lot of space, this method might be called multiple number of times, allocating a new `MasCutoff` everytime. We can do small things like this to reduce (minimal as it may be) the amount of memory we consume during runtime.

Also just as a reminder, all of us collaborators should keep an eye out for leaks like this in each other's PRs, to avoid making the situation worse from newly added code. Pretty sure I am guilty in this regard as well.

Reply:
Thanks for pointing this out. Pushing the change!